account_anglo_saxon_stock_move/model/invoice.py

```diff
@@ -58,65 +58,50 @@ def action_cancel(self):
 
         return super(AccountInvoice, self).action_cancel()
 
-    def reconcile_stock_accrual(self, cr, uid, ids, context=None):
-        context = dict(context or {})
-        ids = isinstance(ids, (int, long)) and [ids] or ids
-        aml_obj = self.pool['account.move.line']
-        amr_obj = self.pool['account.move.reconcile']
-        for inv_brw in self.browse(cr, uid, ids, context=context):
-            for ail_brw in inv_brw.invoice_line:
-                ail_brw.refresh()
-                if not ail_brw.move_id:
-                    continue
-                if ail_brw.move_id.product_id != ail_brw.product_id:
+    @api.multi
+    def reconcile_stock_accrual(self):
+        aml_obj = self.env['account.move.line']
+        for inv_brw in self:
+            # We just care about lines which have a stock_move related
+            move_ids = inv_brw.invoice_line.mapped('move_id')
+
+            categ_ids = move_ids.mapped('product_id.categ_id').filtered(
+                lambda c: not c.property_stock_journal)
+            if categ_ids:
+                raise ValidationError(_(
+                    'The Stock Journal is missing on following '
+                    'product categories: %s' % (', '.join(
+                        categ_ids.mapped('name')))
+                ))
+
+            res = {}
+            # Only stack those that are fully reconciled
+            amr_ids = move_ids.mapped('aml_ids.reconcile_id')
+            amr_ids.unlink()
+
+            for aml_brw in aml_obj.search([('sm_id', 'in', move_ids.ids)]):
+                sm_id = aml_brw.sm_id
+                account_id = aml_brw.account_id.id
+                res.setdefault((sm_id, account_id), aml_obj)
```

Comment:
`setdefault` returns the value if exists then you could use directly since that is a object used with pointers, I mean

```python
value = res.setdefault((sm_id, account_id), aml_obj)
value |= aml_brw
```

This avoid double indexed
<img width="750" alt="screen shot 2017-08-08 at 23 23 43" src="https://user-images.githubusercontent.com/6644187/29104843-8bd9c018-7c8e-11e7-844c-e69b422418f8.png">